src/runtime/graph/graph_runtime.cc

```diff
@@ -433,47 +46,88 @@ void GraphRuntime::LoadParams(dmlc::Stream* strm) {
   }
 }
 
+// Return storage id to device type map. This map will be used to help memory
+// allocation for the storage pool of each device. It will be also used to
+// allocate memory to each data_entry_.
+StorageDeviceMap GraphRuntime::GetStorageDeviceMap() const {
+  StorageDeviceMap sid_dev_map;
+  for (uint32_t nid = 0; nid < this->num_nodes(); ++nid) {
+    const auto &inode = nodes_[nid];
+    for (const auto &e : inode.inputs) {
+      uint32_t eid = this->entry_id(e);
+      uint32_t sid = attrs_.storage_id[eid];
+      sid_dev_map[sid] = nodes_[e.node_id].device;
+    }
+  }
+  // Get all output entries.
+  for (const auto& output : outputs_) {
+    uint32_t eid = this->entry_id(output);
+    uint32_t sid = attrs_.storage_id[eid];
+    sid_dev_map[sid] = nodes_[output.node_id].device;
+  }
+  return sid_dev_map;
+}
+
 void GraphRuntime::SetupStorage() {
   // Grab saved optimization plan from graph.
   std::vector<TVMType> vtype;
   for (const std::string& s_type : attrs_.dltype) {
     vtype.push_back(tvm::runtime::String2TVMType(s_type));
   }
-  data_entry_.resize(num_node_entries());
-  // size of each storage pool entry
-  std::vector<size_t> pool_entry_bytes;
-  // Find the maximum space size.
+
+  StorageDeviceMap sid_dev_map = GetStorageDeviceMap();
+  std::unordered_map<DLDeviceType, std::unordered_map<size_t, size_t>,
+                     DLDeviceTypeHash>
+      device_pool_entry_bytes;
+
+  // Find the maximum space size for each device.
   for (size_t i = 0; i < attrs_.shape.size(); ++i) {
-    int storage_id = attrs_.storage_id[i];
+    uint32_t sid = static_cast<uint32_t>(attrs_.storage_id[i]);
     size_t size = 1;
     for (int64_t sz : attrs_.shape[i]) {
       size *= static_cast<size_t>(sz);
     }
-    CHECK_GE(storage_id, 0) << "Do not support runtime shape op";
+    CHECK_GE(sid, 0) << "Do not support runtime shape op";
     DLDataType t = vtype[i];
     size_t bits = t.bits * t.lanes;
     CHECK_EQ(bits % 8U, 0U);
     size_t bytes = (bits / 8U) * size;
 
-    size_t sid = static_cast<size_t>(storage_id);
-    if (sid >= pool_entry_bytes.size()) {
-      pool_entry_bytes.resize(sid + 1, 0);
-    }
-    pool_entry_bytes[sid] = std::max(pool_entry_bytes[sid], bytes);
+    DLDeviceType dev_type = sid_dev_map[sid];
+    device_pool_entry_bytes[dev_type][sid] =
```

Comment:
looks like there's no need to make `dev_type` a key. it can be achieved from `sid_dev_map` right? just feel better to keep data struct simple.

Reply:
Good point. Thanks.